### PR TITLE
Finalize trigger enables propagation if it was stopped

### PR DESCRIPTION
**Pull request by prolic**
superseeds https://github.com/prooph/service-bus/pull/183

/cc @UFOMelkor

### CODE REVIEW

**Comment by UFOMelkor** [comment]
Please give me 2 hours to evaluate another test case that came to my mind.

**Comment by UFOMelkor** [comment]
If I understand the logic of the finalize event correct, It's listeners should be always called. This would not be the case if the propagation is stopped but no exception has been thrown, only the first finalize listener is called:

```php
    /**
     * @test
     */
    public function it_always_triggers_finalize_listeners_regardless_whether_the_propagation_of_the_event_has_been_stopped(): void
    {
        $messageBus = new CommandBus();

        $messageBus->attach(CommandBus::EVENT_DISPATCH, function (ActionEvent $event) {
            $event->setParam(CommandBus::EVENT_PARAM_MESSAGE_HANDLER, function () {
            });
        }, CommandBus::PRIORITY_LOCATE_HANDLER+1);
        $messageBus->attach(CommandBus::EVENT_DISPATCH, function (ActionEvent $event) {
            $event->stopPropagation();
        }, CommandBus::PRIORITY_INVOKE_HANDLER-1);

        $messageBus->attach(MessageBus::EVENT_FINALIZE, function () {
        }, 3);
        $finalizeHasBeenCalled = false;
        $messageBus->attach(MessageBus::EVENT_FINALIZE, function () use (&$finalizeHasBeenCalled) {
            $finalizeHasBeenCalled = true;
        }, 2);

        $messageBus->dispatch('a message');

        self::assertTrue($finalizeHasBeenCalled);
    }
```

If my assumption is correct, additional tests for the CommandBus, EventBus and QueryBus are necessary and the [`triggerFinalize`](https://github.com/prooph/service-bus/blob/a9b0a7e9d9ffb35cd7a716a552e543be177e171f/src/MessageBus.php#L102) method should be like
```php
protected function triggerFinalize(ActionEvent $actionEvent): void
    {
        $actionEvent->setName(self::EVENT_FINALIZE);
        $actionEvent->stopPropagation(false);

        $this->events->dispatch($actionEvent);
    }
```

**Comment by prolic** [comment]
@UFOMelkor I added further tests

File: tests/QueryBusTest.php [hunk 1]
```diff
@@ -352,4 +352,30 @@ function (ActionEvent $actionEvent) use (&$exceptionParamWasSet): void {
...
+     */
+    public function it_always_triggers_finalize_listeners_regardless_whether_the_propagation_of_the_event_has_been_stopped(): void
+    {
+        $this->queryBus->attach(QueryBus::EVENT_DISPATCH, function (ActionEvent $event) {
+            $event->setParam(QueryBus::EVENT_PARAM_MESSAGE_HANDLER, function (): void {
+            });
+        }, QueryBus::PRIORITY_LOCATE_HANDLER + 1);
+        $this->queryBus->attach(QueryBus::EVENT_DISPATCH, function (ActionEvent $event): void {
+            $event->stopPropagation();
+            throw new \RuntimeException('boom');
```
Comment (UFOMelkor) [inline tests/QueryBusTest.php]: 3

File: tests/EventBusTest.php [hunk 2]
```diff
@@ -305,4 +305,34 @@ function (ActionEvent $actionEvent) use (&$finalizeIsTriggered, &$listenerExcept
...
+     */
+    public function it_always_triggers_finalize_listeners_regardless_whether_the_propagation_of_the_event_has_been_stopped(): void
+    {
+        $this->eventBus->attach(EventBus::EVENT_DISPATCH, function (ActionEvent $event) {
+            $event->setParam(EventBus::EVENT_PARAM_MESSAGE_HANDLER, function (): void {
+            });
+        }, EventBus::PRIORITY_LOCATE_HANDLER + 1);
+        $this->eventBus->attach(EventBus::EVENT_DISPATCH, function (ActionEvent $event): void {
+            $event->stopPropagation();
+            throw new \RuntimeException('boom');
```
Comment (UFOMelkor) [inline tests/EventBusTest.php]: 2

File: tests/CommandBusTest.php [hunk 3]
```diff
@@ -295,4 +295,34 @@ public function it_passes_queued_commands_to_command_dispatch_exception_in_case_
...
+     */
+    public function it_always_triggers_finalize_listeners_regardless_whether_the_propagation_of_the_event_has_been_stopped(): void
+    {
+        $this->commandBus->attach(CommandBus::EVENT_DISPATCH, function (ActionEvent $event) {
+            $event->setParam(CommandBus::EVENT_PARAM_MESSAGE_HANDLER, function (): void {
+            });
+        }, CommandBus::PRIORITY_LOCATE_HANDLER + 1);
+        $this->commandBus->attach(CommandBus::EVENT_DISPATCH, function (ActionEvent $event): void {
+            $event->stopPropagation();
+            throw new \RuntimeException('boom');
```
Comment (UFOMelkor) [inline tests/CommandBusTest.php]: 1

**Comment by prolic** [comment]
@UFOMelkor You're right, fixed!

**Comment by UFOMelkor** [comment]
LGTM 👍

**Pull request merged**